[ZT] AI wrapper tenant control tutorial (#21756)
cloudflare/cloudflare-docs: src/content/docs/cloudflare-one/tutorials/ai-wrapper-tenant-control.mdx

---

> updated: 2025-04-25

Author: ```suggestion
updated: 2025-05-02
```

</TabItem> <TabItem label="Dashboard">

1. Log in to the [Cloudflare dashboard](https://dash.cloudflare.com/) and select your account.
2. Go to **Workers & Pages** > **Workers & Pages**.
3. Select **Create**.
4. In **Workers**, choose the **Hello world** template.
5. Name your worker, then select **Deploy**.
6. Select your Worker, then go to the **Settings** tab.
7. Go to **Variables and Secrets**, then select **Add**.
8. Choose _Secret_ as the type, name your secret (for example, `OPENAI_API_KEY`), and enter the value of your AI provider's API key in **Value**.

You can now build the Worker using the online code editor by selecting **Edit code** on your Worker page.

</TabItem> </Tabs>

### 2. Build the Worker

The following is an example starter Worker that serves a simple front-end to allow a user to interact with an AI provider behind AI Gateway. This example uses OpenAI as its AI provider:

<Details header="Example AI wrapper Worker">

```javascript
export default {
	async fetch(request, env) {
		if (request.url.endsWith("/api/chat")) {
			if (request.method === "POST") {
				try {
					const { messages } = await request.json();

					const response = await fetch(

> 						"https://gateway.ai.cloudflare.com/v1/<ACCOUNT_ID>/<GATEWAY_ID>/openai/chat/completions",

Reviewer: Nit: We could highlight this line, since we mention it below (it contains two placeholders that must be replaced).

              messagesDiv.appendChild(messageDiv);
              messagesDiv.scrollTop = messagesDiv.scrollHeight;
          }
      </script>
  </body>
  </html>`;
```

</Details>

Note that the account ID and gateway ID need to be replaced in the AI Gateway endpoint. You can add these as [environment variables](/workers/configuration/environment-variables/) or [secrets](/workers/configuration/secrets/) in Workers. If you chose to use Authenticated Gateway when creating your AI gateway, make sure to also add your token as a secret and pass its value to the AI gateway in the `cf-aig-authorization` header.

### 3. Publish the Worker

Once the Worker code is complete, you need to make the Worker addressable using a hostname controllable by Cloudflare Access.

<Tabs> <TabItem label="Wrangler">

Edit the `wrangler.toml` configuration file and add the following information to ensure that the worker is only accessible using the custom hostname:

```toml
name = "ai-agent-wrapper"
main = "src/index.js"
compatibility_date = "2023-10-30"
workers_dev = false

# replace with your custom domain
routes = [
  { pattern = "<YOUR_CUSTOM_DOMAIN>", custom_domain = true }

> ]

Reviewer: Nit: We could highlight these lines that have been added to the file.

[vars]
# Add any environment variables here
```

To publish the worker, run `wrangler publish`.

</TabItem> <TabItem label="Dashboard">

If the worker was built remotely using the online code editor available in the **Dash**, it can be published by clicking **Deploy**.

> In order to ensure that the worker is only accessible via the custom hostname, do:

Reviewer: ```suggestion
To ensure that the worker is only accessible via the custom hostname, do the following:
```

> 2. Go to **Workers & Pages** > **Workers & Pages**.
> 3. Select your Worker.
> 4. Go to **Settings**.
> 5. Within **Domains & Routes**, select **Add**.
> 6. Choose **Custom domain**.
> 7. Enter your desired custom domain name.
> 8. Select **Add domain**.

Reviewer: ```suggestion
1. Go to **Workers & Pages** > **Workers & Pages**.
2. Select your Worker.
3. Go to **Settings**.
4. Within **Domains & Routes**, select **Add**.
5. Choose **Custom domain**.
6. Enter your desired custom domain name.
7. Select **Add domain**.
```
Renumbered the steps.
